collectd/collectd.go

```diff
@@ -0,0 +1,186 @@
+}
+
+type Server struct {
+	mu sync.Mutex
+	wg sync.WaitGroup
+
+	conn *net.UDPConn
+
+	writer      SeriesWriter
+	Database    string
+	typesdb     gollectd.Types
+	typesdbpath string
+}
+
+func NewServer(w SeriesWriter, typesDBPath string) *Server {
+	s := Server{
+		writer:      w,
+		typesdbpath: typesDBPath,
+		typesdb:     make(gollectd.Types),
+	}
+
+	return &s
+}
+
+func ListenAndServe(s *Server, iface string) error {
+	if iface == "" { // Make sure we have an address
+		return errors.New("bind address required")
+	} else if s.Database == "" { // Make sure they have a database
+		return errors.New("database was not specified in config")
+	}
+
+	addr, err := net.ResolveUDPAddr("udp", iface)
+	if err != nil {
+		return fmt.Errorf("unable to resolve UDP address: %v", err)
+	}
+
+	s.typesdb, err = gollectd.TypesDBFile(s.typesdbpath)
+	if err != nil {
+		return fmt.Errorf("unable to parse typesDBFile: %v", err)
+	}
+
+	conn, err := net.ListenUDP("udp", addr)
+	if err != nil {
+		return fmt.Errorf("unable to listen on UDP: %v", err)
+	}
+	s.conn = conn
+
+	s.wg.Add(1)
+	go s.serve(conn)
+
+	return nil
+}
+
+func (s *Server) serve(conn *net.UDPConn) {
+	defer s.wg.Done()
+
+	// From https://collectd.org/wiki/index.php/Binary_protocol
+	//   1024 bytes (payload only, not including UDP / IP headers)
+	//   In versions 4.0 through 4.7, the receive buffer has a fixed size
+	//   of 1024 bytes. When longer packets are received, the trailing data
+	//   is simply ignored. Since version 4.8, the buffer size can be
+	//   configured. Version 5.0 will increase the default buffer size to
+	//   1452 bytes (the maximum payload size when using UDP/IPv6 over
+	//   Ethernet).
+	buffer := make([]byte, 1452)
+
+	for {
+		n, _, err := conn.ReadFromUDP(buffer)
+		if err != nil {
+			if s.conn == nil {
+				// This error occured becase we closed the connection and this is expected behavior
+				return
+			}
+			log.Printf("Collectd ReadFromUDP error: %s", err)
+		}
+
+		// Read in data in a separate goroutine.
+		s.handleMessage(buffer[:n])
+	}
+}
+
+func (s *Server) handleMessage(buffer []byte) {
+	packets, err := gollectd.Packets(buffer, s.typesdb)
+	if err != nil {
+		log.Printf("Collectd parse error: %s", err)
+		return
+	}
+
+	for _, packet := range *packets {
+		metrics := Unmarshal(&packet)
+		for _, m := range metrics {
+			// Convert metric to a field value.
+			var values = make(map[string]interface{})
+			values[m.Name] = m.Value
+
+			err := s.writer.WriteSeries(s.Database, "", m.Name, m.Tags, m.Timestamp, values)
+			if err != nil {
+				log.Printf("Collectd cannot write data: %s", err)
+				continue
+			}
+		}
+	}
+}
+
+// Close shuts down the server's listeners.
+func (s *Server) Close() error {
+	// Notify other goroutines of shutdown.
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if s.conn == nil {
+		return errors.New("server already closed")
```

Comment:
That's a fair point. I generally like the "well-definedness" of declared errors but you're right that it does expose more through the API.